subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/coreTypes/GenericMapTypeHandler.java

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.persistence.typeHandling.coreTypes;
+
+import com.google.common.collect.Maps;
+import org.terasology.persistence.typeHandling.PersistedData;
+import org.terasology.persistence.typeHandling.PersistedDataSerializer;
+import org.terasology.persistence.typeHandling.TypeHandler;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class GenericMapTypeHandler<K, V> extends TypeHandler<Map<K, V>> {
+
+    private static final String KEY = "key";
+    private static final String VALUE = "value";
+
+    private final TypeHandler<K> keyHandler;
+    private final TypeHandler<V> valueHandler;
+
+    public GenericMapTypeHandler(TypeHandler<K> keyHandler, TypeHandler<V> valueHandler) {
+        this.keyHandler = keyHandler;
+        this.valueHandler = valueHandler;
+    }
+
+    @Override
+    protected PersistedData serializeNonNull(Map<K, V> data, PersistedDataSerializer serializer) {
+        final List<PersistedData> entries = data.entrySet().stream()
+                .map(entry -> serializeEntry(entry, serializer))
+                .collect(Collectors.toList());
+        return serializer.serialize(entries);
+    }
+
+    private PersistedData serializeEntry(Map.Entry<K, V> entry, PersistedDataSerializer serializer) {
+        PersistedData key = keyHandler.serialize(entry.getKey(), serializer);
+        PersistedData value = valueHandler.serialize(entry.getValue(), serializer);
+
+        Map<String, PersistedData> result = Maps.newLinkedHashMap();
+        if (!key.isNull()) {
+            result.put(KEY, key);
+            result.put(VALUE, value);
+        }
+        return serializer.serialize(result);
+    }
+
+    @Override
+    public Optional<Map<K, V>> deserialize(PersistedData data) {
+        if (!data.isArray()) {
+            return Optional.empty();
+        }
+
+        Map<K, V> result = Maps.newLinkedHashMap();
+
+        for (PersistedData entry : data.getAsArray()) {
+            final Optional<K> key = keyHandler.deserialize(entry.getAsValueMap().get(KEY));
+            final Optional<V> value = valueHandler.deserialize(entry.getAsValueMap().get(VALUE));
+
+            key.ifPresent(k -> result.put(k, value.orElse(null)));
```

Comment:
Also a good question (which I think holds for other type handlers as well...). The nested nature of the maps and other collections yields two-fold results: the whole thing cannot be serialized (`Optional.empty()`) vs only parts cannot be serialized. 
Maybe we should print a warning in the type handler if that happens. What kind of data would we need in that case?

Reply:
I guess for deserialization there are times when you would want to do a forgiving lossy-loading thing like this. Maybe the server admin removed a module and you want to take your chances with just dropping that part of the data structure while loading the rest of it.

(I haven't yet learned which use cases we're talking about here. Is this for configuration files _and_ save files _and_ the wire protocol?)

There are other times when I'd want a strict-mode, when there's no _good_ reason for anything to fail to deserialize and it might as declare the entire operation invalid.

For debugging, you want the raw key and value as well as some way to _find_ it if the key name isn't very descriptive and it's nested somewhere deep. I don't think `deserialize` is being given that context here so that kind of location information might be beyond the scope of this PR.

Reply:
@keturn that is actually something I was thinking about during the initial serialization overhaul, and I came to the same conclusions as you. The reason I didn't implement a strict-mode was because at the time I deemed it to be too much effort for its benefits. Definitely would be nice to have though -- maybe even just as a standalone "prefab validator".

@skaldarnar I would definitely log a warning/error here -- that is what all other type handlers also do when they cannot find something that they need. Something like `"Missing '" + KEY + "' field"` if the `KEY` field is missing (and the same if the `VALUE` field is absent). We do not *have* to log a warning if there are other redundant fields present (other type handlers do not), but you could if you wanted to.